enhancements/machine-api/control-plane.md

```diff
@@ -0,0 +1,281 @@
+
+There are multiple components and operators involved in the lifecycle management of the Control Plane. This in addition to its complex nature makes it difficult to predict the holistic behaviour for failure scenarios. To mitigate risks the Control Plane controller will short-circuit preventing any scaling operation until its healthiness criteria is met again.
+
+## Design Details
+
+### Test Plan
+
+- Exhaustive unit testing.
+- Exhaustive integration testing via [envTest](https://book.kubebuilder.io/reference/testing/envtest.html).
+- E2e testing on the machine API [e2e test suite](https://github.com/openshift/cluster-api-actuator-pkg/tree/master/pkg):
+	- Scale from 3 to 5.
+	- Scale from 5 to 3.
+	- Vertical scaling Rolling upgrade.
+	- Simulate 1 unhealthy node and remediation.
+- E2e testing on the origin e2e test suite.
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+This controller will go GA in 4.N as long as:
+- All the testing above is in place.
+- There is considerable internal usage and manual disruptive QE testing feedback.
+- This is beta tested pre-release during a considerable period of time either internally or by an early adopter user through the dev preview builds.
+
+Otherwise this would go "Tech Preview" under a feature gate first until enough real use feedback is developed. The risk though is that there might be a very scarce use of feature gated features in the field as to develop the feedback we need.
+
+### Upgrade / Downgrade Strategy
+
+During the cluster upgrade for the targeted release the Machine API Operator (MAO) will let the CVO to instantiate the new CRD `controlPlane` and it will run the backing controller making this functionallity opt-in for existing clusters. The user can create an instance of the new CRD if they choose to do so.
+
+New IPI clusters deployed after the targeted release will run the `controlPlane` instance deployed by the installer out of the box.
+
+### Version Skew Strategy
+
+## Implementation History
+- "https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20191017-kubeadm-based-control-plane.md"
+- "https://github.com/openshift/enhancements/blob/master/enhancements/etcd/cluster-etcd-operator.md"
+- "https://github.com/openshift/machine-config-operator/blob/master/docs/etcd-quorum-guard.md"
+
+## Drawbacks
+
+## Alternatives
+
+- Same approach but different details:
+The Control Plane controller scaling logic could be built atop machineSets just like a deployment uses replica sets. The scenarios to be supported in this proposal are very specific. The flexibility that managing machineSets would provide is not needed. The Control Plane is critical and we want to favour control over flexibility here, therefore this proposes an ad-hoc controller logic to avoid unnecessary layers of complexity.
```

Comment:
line break every 120 or every sentence (whichever is shorter) to make commenting easier.

Can you expand on why we would not re-use this existing metaphor with some admission webhook validators to restrict how far the adjustments can go?  This seems be the "standard" API that existing for scaling up and down and I would have expected us to double-down on the existing metaphor instead of creating something new and special.

Reply:
Aside from auto-scaling - which it seems isn't the goal of the proposal as it stands, although it makes sense to me to add it - what would be the main function of this controller if it was built on machinesets?

Say for scale-up, would this come down to choose-which-control-plane-machineset-to-scale-up and an `(all Etcd members are healthy AND all owned machines have a backed ready node)` safety wrapper?

Reply:
> Say for scale-up, would this come down to choose-which-control-plane-machineset-to-scale-up and an `(all Etcd members are healthy AND all owned machines have a backed ready node)` safety wrapper?

For this scale-up case, could we add additional extra layers of safety around cluster-etcd-operator's decision to add a new etcd member, such that it's safe to add new control plane compute capacity and have cluster-etcd-operator worry about the etcd operational aspects?

Reply:
If we could drop that safety check and assume scale compute in/out is always completely safe this controller would come down to:
- Ensure even spread of compute resources across failure domains and managing the underlying MHC.

I'd love that but not sure that'd be ever feasible for all scenarios though. Say e.g for scaling in when 1 etcd member out of 3 is unhealthy a "cattle" scale down can easily break quorum. I tried to elaborate more my etcd concerning scenarios in `alternatives` and `proposal`

Reply:
cc @hexfusion for feedback on the etcd scenarios and how to possibly decouple more compute from etcd healthiness

Reply:
If we can make sure any concerning scenario as in https://github.com/openshift/enhancements/pull/278/files#diff-dcd7423e408e8b740b1b9b14e53eeda1R293-R316 would be handled gracefully by etcd operator then we can drop the any etcd “pet” logic from this proposal and treat this as regular compute scaling operations.